Verify doc examples

grahamc opened the pull request:
###### Motivation for this change

Test the function examples actually work.


Testing steps:

1. enter doc/
2. nix-shell
3. make doc-tests


It outputs something like:

```
$ make doc-tests 
cd ./functions/library; python3 ./test.py
function-library-lib.attrset.attrByPath-example-value-exists: pass
function-library-lib.attrset.attrByPath-example-default-value: pass
function-library-lib.attrsets.recursiveUpdateUntil-example: pass
function-library-lib.attrsets.recursiveUpdate-example: pass
function-library-lib.attrsets.matchAttrs-example-matching-pattern-deep: pass
5 passed, 0 failed

```

For each example without a testable program, it prints:

```
Skipping example function-library-lib.attrsets.zipAttrs-example: no <programlisting role="input">
```

If a test fails, it outputs instructions on how to test only the failing examples:

```
function-library-lib.attrsets.recursiveUpdateUntil-example: pass
function-library-lib.attrsets.recursiveUpdate-example: pass
--------  Doc test failure  --------
ID: function-library-lib.attrsets.matchAttrs-example-matching-pattern-shallow
Executing:


      {
        actual = (let lib = import /home/grahamc/projects/nixpkgs/lib; in (
          
lib.attrsets.matchAttrs
  { cpu = {}; }
  { cpu = { bits = 64; }; }

        ));
        expect = (
          
  false

        );
      }
    

Expecting:

  false

Found:
True
----
--------  Doc test failure  --------
ID: function-library-lib.attrsets.matchAttrs-example-matching-pattern-deep
Executing:


      {
        actual = (let lib = import /home/grahamc/projects/nixpkgs/lib; in (
          
lib.attrsets.matchAttrs
  { host = { cpu = {}; }; }
  { host = { cpu = { bits = 64; }; }; }

        ));
        expect = (
          false
        );
      }
    

Expecting:
false
Found:
True
----
4 passed, 2 failed
Retry failing examples with:
./functions/library/test.py --lib-dir /home/grahamc/projects/nixpkgs/lib function-library-lib.attrsets.matchAttrs-example-matching-pattern-shallow function-library-lib.attrsets.matchAttrs-example-matching-pattern-deep
```

These tests will block the Nixpkgs manual from building as well.

The XML isn't super nice to write, but I think isn't too onerous?

Anyway, rendered it looks like this: 

![image](https://user-images.githubusercontent.com/76716/46988078-ee996d80-d0c4-11e8-8a69-7f0ee69c4729.png)

I also verified that python3 was already in the build closure.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

grahamc committed:
nixpkgs docs: add doc tests

Also, hide <programlisting role="context"> which are for doc-tests.

grahamc committed:
nixpkgs doc: add doctests to some attrsets functions

grahamc committed:
nixpkgs docs: document matchAttrs

grahamc committed:
nixpkgs docs: document overrideExisting

FRidh commented on doc/functions/library/test.py:
```diff
@@ -0,0 +1,124 @@
+        yield { "id": xml_id, "code": code.text, "expect": expect.text }
+
+
+parser = argparse.ArgumentParser(description='Execute doc tests for the Nixpkgs manual..')
```
put the following code in a function, e.g. `main()` and add `if __name __ = "__main__": main()

FRidh commented on doc/functions/library/test.py:
```diff
@@ -0,0 +1,124 @@
+doc_root = os.path.dirname(os.path.realpath(__file__))
+all_doc_files = glob.glob("{}/*.xml".format(doc_root))
+
+all_docs = [ET.parse(filename) for filename in all_doc_files]
```
generator expression

graham-at-target replied:
Hmm I'm not sure what to do with this feedback?

FRidh commented on doc/functions/library/test.py:
```diff
@@ -0,0 +1,124 @@
+all_docs = [ET.parse(filename) for filename in all_doc_files]
+all_examples = itertools.chain(*[find_examples(doc) for doc in all_docs])
+
+if len(args.ids) > 1:
```
if args.ids:

shlevy replied:
I don't think so, this is >1 not >0

shlevy replied:
... But should it be? seems like we drop a singleton list here? OK agreed with @FRidh then

graham-at-target replied:
+1 this is a relic of when I didn't have proper argument parsing 😁 good catch.

FRidh commented on doc/default.nix:
```diff
@@ -6,7 +6,8 @@ in
 
-  buildInputs = with pkgs; [ pandoc libxml2 libxslt zip jing  xmlformat ];
+  buildInputs = with pkgs; [ pandoc libxml2 libxslt zip jing xmlformat
+    python36 nix ];
```
python3 is sufficient. Note that soon I intend to make python37 the default, and this code will still work fine then.

shlevy commented on doc/functions/library/test.py:
```diff
@@ -0,0 +1,116 @@
+failed = []
+
+for example in filtered_examples:
+    full_expression = """
```
Maybe want [f strings](https://www.python.org/dev/peps/pep-0498/)?

shlevy reviewed (approved):
Nice!

grahamc committed:
nixpkgs docs: Extract XSL parameters in to a common XSL file

Using a single configuration file reduces the need to think about
how quoting works from Nix -> Bash -> Make -> xsltproc -> XSL

Additionally, it allows us to make additional interesting
customisations as needed.

grahamc committed:
nixpkgs doc: document getOutput

grahamc committed:
fixup line lengths, add output diffs

grahamc committed:
Document getDev, getLib, getBin

grahamc committed:
a few more attrset functions

mmahut commented:
Are there any updates on this pull request, please?

asymmetric commented:
@grahamc has this been abandoned?
